include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,110 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_next_event_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_NEXT_EVENT_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + THREAD_EVENT_MIN_WAIT)
+
+void thread_event_assert_invariants_debug(tsd_t *tsd);
+void thread_event_trigger(tsd_t *tsd, uint64_t usize, bool delay_event);
+void thread_event_rollback(tsd_t *tsd, size_t diff);
+void thread_event_update(tsd_t *tsd);
+
+/*
+ * List of all event wait times, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
```

Comment:
nit: `ITERATE_OVER_ALL_EVENTS`

Reply:
Makes sense. Also changed the `W` macro to `E` to be consistent.